modules/renter/contractor/contractor.go

```diff
@@ -81,41 +76,76 @@ func (c *Contractor) SetAllowance(a modules.Allowance) error {
 		return errors.New("renew window must be less than period")
 	}
 
-	err := c.formContracts(a)
+	// check that allowance is sufficient to store at least one sector
+	numSectors, err := maxSectors(a, c.hdb)
 	if err != nil {
 		return err
+	} else if numSectors == 0 {
+		return errInsufficientAllowance
 	}
 
-	// Set the allowance.
+	// prevent ProcessConsensusChange from renewing contracts until we have
+	// finished
+	c.contractLock.Lock()
+	defer c.contractLock.Unlock()
+
 	c.mu.Lock()
-	c.allowance = a
-	err = c.saveSync()
+	endHeight := c.blockHeight + a.Period
+
+	// check if allowance has different period or funds; if so, we should
+	// renew existing contracts with the new allowance.
+	// TODO: compare numSectors instead of allowance.Funds?
+	var renewSet []modules.RenterContract
+	if a.Period != c.allowance.Period || a.Funds.Cmp(c.allowance.Funds) != 0 {
+		for _, contract := range c.contracts {
+			renewSet = append(renewSet, contract)
+		}
+		// TODO: dangerous -- figure out a safer approach later
+		c.contracts = map[types.FileContractID]modules.RenterContract{}
```

Comment:
The problem is that in the meantime, people are calling `c.Contracts()` and `ProcessConsensusChange` may be trying to auto-renew. So the old contracts cannot be left in `c.contracts` while `SetAllowance` is running. They have to be atomically replaced.
This suggests an alternative, in which you would incrementally renew contracts and store them in some `tempContracts` map, and then atomically swap `tempContracts` for `c.contracts`. But this is even worse, because if you exit early you lose all of the renewed contracts.
To rectify this, you could move `tempContracts` into the contractor and save it to disk. But there are still problems! Namely, until you do the atomic swap, `ProcessConsensusChange` will still be looking at the old contracts, and thus it may try to renew them, which is wasteful at best.
So yes, lots of edge cases :P

This brings up a good question: what happens when you try to renew a contract and revise it at the same time? Will the host reject one of the operations? This scenario will occur if the user sets a new allowance while files are being uploaded.